File: src/app/api/convert/route.ts
import { NextResponse } from "next/server";
import { writeFile, unlink } from "fs/promises";
import { convertToMarkdown } from "~/server/converter";
import path from "path";
import os from "os";

interface ConvertRequest {
  fileUrl: string;
}

export async function POST(req: Request) {
  let tempFilePath: string | undefined;

  try {
    const data = (await req.json()) as ConvertRequest;

    if (!data.fileUrl) {
      return NextResponse.json(
        { error: "No file URL provided" },
        { status: 400 },
      );
    }

    // Download the file from the URL
    const response = await fetch(data.fileUrl);
    if (!response.ok) {
      return NextResponse.json(
        { error: "Failed to download file" },
        { status: 500 },
      );
    }

    // Create a temporary file path using os.tmpdir()
    tempFilePath = path.join(
      os.tmpdir(),
      `${Date.now()}-${Math.random().toString(36).substring(7)}`,
    );

    // Save the file
    const buffer = await response.arrayBuffer();
    await writeFile(tempFilePath, Buffer.from(buffer));

    // Convert the file
    const result = await convertToMarkdown(tempFilePath);

    if (!result.success) {
      return NextResponse.json({ error: result.error }, { status: 500 });
    }

    return NextResponse.json({ markdown: result.markdown });
  } catch (error) {
    console.error("Conversion error:", error);
    return NextResponse.json(
      { error: "Internal server error" },
      { status: 500 },
    );
  } finally {
    // Clean up the temporary file
    if (tempFilePath) {
      try {
        await unlink(tempFilePath);
      } catch (error) {
        console.error("Failed to clean up temporary file:", error);
      }
    }
  }
}

File: src/app/api/convert/route.py
from http.server import BaseHTTPRequestHandler
import json
import os
import tempfile
from urllib.request import urlopen
from markitdown import convert_to_markdown

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        # Read the request body
        content_length = int(self.headers['Content-Length'])
        body = self.rfile.read(content_length)
        data = json.loads(body)
        
        file_url = data.get('fileUrl')
        if not file_url:
            self.send_error(400, "No file URL provided")
            return

        try:
            # Download the file
            response = urlopen(file_url)
            
            # Create a temporary file
            with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
                tmp_file.write(response.read())
                tmp_path = tmp_file.name

            try:
                # Convert the file
                markdown = convert_to_markdown(tmp_path)
                
                # Send response
                self.send_response(200)
                self.send_header('Content-type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({
                    "success": True,
                    "markdown": markdown,
                    "error": None
                }).encode('utf-8'))
                
            finally:
                # Clean up
                os.unlink(tmp_path)
                
        except Exception as e:
            self.send_response(500)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({
                "success": False,
                "markdown": None,
                "error": str(e)
            }).encode('utf-8')) 